File: scripts/artifact_identity.py
#!/usr/bin/env python3
"""Generate a secret-safe identity report for final Basecamp release artifacts."""

from __future__ import annotations

import argparse
import hashlib
import json
import os
from pathlib import Path, PurePosixPath
import re
import tarfile
import tempfile
import sys

VARIANT = "linux-amd64"
MAX_FILES = 2048
MAX_FILE_BYTES = 512 * 1024 * 1024


def sha256_file(path: Path) -> tuple[str, int]:
    if path.is_symlink() or not path.is_file():
        raise ValueError(f"artifact is missing, non-file, or symlinked: {path}")
    digest = hashlib.sha256()
    size = 0
    with path.open("rb") as handle:
        while chunk := handle.read(1024 * 1024):
            size += len(chunk)
            digest.update(chunk)
    return digest.hexdigest(), size


def hash_bytes(data: bytes) -> dict[str, object]:
    return {"sha256": hashlib.sha256(data).hexdigest(), "size": len(data)}


def package_payload(archive: Path) -> dict[str, dict[str, object]]:
    archive_hash, _ = sha256_file(archive)
    del archive_hash
    payload: dict[str, dict[str, object]] = {}
    count = 0
    with tarfile.open(archive, "r:gz") as package:
        for member in package:
            name = member.name
            while name.startswith("./"):
                name = name[2:]
            if not name or member.isdir():
                continue
            count += 1
            if count > MAX_FILES:
                raise ValueError(f"too many files in {archive.name}")
            path = PurePosixPath(name)
            if path.is_absolute() or ".." in path.parts or member.issym() or member.islnk() \
                    or not member.isfile():
                raise ValueError(f"unsafe package entry: {member.name}")
            if member.size < 0 or member.size > MAX_FILE_BYTES:
                raise ValueError(f"oversized package entry: {member.name}")
            if path == PurePosixPath("manifest.json"):
                installed_name = "manifest.json"
            elif len(path.parts) >= 3 and path.parts[:2] == ("variants", VARIANT):
                installed_name = PurePosixPath(*path.parts[2:]).as_posix()
            else:
                raise ValueError(f"unexpected package entry: {member.name}")
            if not installed_name or installed_name in payload:
                raise ValueError(f"duplicate package payload path: {installed_name}")
            source = package.extractfile(member)
            if source is None:
                raise ValueError(f"cannot read package entry: {member.name}")
            digest = hashlib.sha256()
            read_size = 0
            while chunk := source.read(1024 * 1024):
                read_size += len(chunk)
                if read_size > member.size or read_size > MAX_FILE_BYTES:
                    raise ValueError(f"package entry exceeds declared size: {member.name}")
                digest.update(chunk)
            if read_size != member.size:
                raise ValueError(f"truncated package entry: {member.name}")
            payload[installed_name] = {"sha256": digest.hexdigest(), "size": read_size}
    payload["variant"] = hash_bytes(VARIANT.encode())
    return payload


def installed_payload(root: Path, ignored_top_levels: frozenset[str] = frozenset()) -> dict[str, dict[str, object]]:
    if root.is_symlink() or not root.is_dir():
        raise ValueError(f"installed component is missing or symlinked: {root}")
    payload: dict[str, dict[str, object]] = {}
    for current, directories, files in os.walk(root, followlinks=False):
        current_path = Path(current)
        if current_path == root:
            directories[:] = [name for name in directories if name not in ignored_top_levels]
        for name in directories:
            path = current_path / name
            if path.is_symlink():
                raise ValueError(f"installed payload contains symlink: {path.relative_to(root)}")
        for name in files:
            path = current_path / name
            if path.is_symlink() or not path.is_file():
                raise ValueError(f"installed payload contains non-file or symlink: {path.relative_to(root)}")
            relative = path.relative_to(root).as_posix()
            digest, size = sha256_file(path)
            payload[relative] = {"sha256": digest, "size": size}
            if len(payload) > MAX_FILES:
                raise ValueError(f"too many installed files in {root.name}")
    return payload


def component_report(archive: Path, installed: Path,
                     ignored_top_levels: frozenset[str] = frozenset()) -> dict[str, object]:
    expected = package_payload(archive)
    actual = installed_payload(installed, ignored_top_levels)
    if expected != actual:
        missing = sorted(set(expected) - set(actual))
        extra = sorted(set(actual) - set(expected))
        changed = sorted(name for name in set(expected) & set(actual) if expected[name] != actual[name])
        raise ValueError(
            f"payload mismatch for {installed.name}: missing={missing}, extra={extra}, changed={changed}"
        )
    return {
        "file_count": len(actual),
        "files": {name: actual[name] for name in sorted(actual)},
    }


def artifact(path: Path) -> dict[str, object]:
    digest, size = sha256_file(path)
    return {"name": path.name, "sha256": digest, "size": size}


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("--ui-lgx", required=True, type=Path)
    parser.add_argument("--core-lgx", required=True, type=Path)
    parser.add_argument("--iso", required=True, type=Path)
    parser.add_argument("--appimage", required=True, type=Path)
    parser.add_argument("--core-rev", required=True)
    parser.add_argument("--output", required=True, type=Path)
    return parser.parse_args()


def main() -> int:
    args = parse_args()
    if re.fullmatch(r"[0-9a-f]{40}", args.core_rev) is None:
        raise ValueError("peers_core revision must be an exact 40-character lowercase commit")
    output = args.output.absolute()
    extra = Path("/extra").resolve()
    if extra not in output.resolve(strict=False).parents:
        raise ValueError("identity report must be written under /extra")

    gui = args.iso.absolute() / "data/Logos/LogosBasecamp"
    report = {
        "schema": 1,
        "peers_core_revision": args.core_rev,
        "artifacts": {
            "basecamp_appimage": artifact(args.appimage),
            "peers_core_lgx": artifact(args.core_lgx),
            "peers_ui_lgx": artifact(args.ui_lgx),
        },
        "components": {
            "peers_core": component_report(args.core_lgx, gui / "modules/peers_core"),
            "peers_ui": component_report(
                args.ui_lgx,
                gui / "plugins/peers_ui",
                frozenset({"media-cache"}),
            ),
        },
    }
    output.parent.mkdir(parents=True, exist_ok=True)
    with tempfile.NamedTemporaryFile("w", dir=output.parent, prefix=f".{output.name}.",
                                     delete=False, encoding="utf-8") as handle:
        temporary = Path(handle.name)
        os.fchmod(handle.fileno(), 0o600)
        json.dump(report, handle, sort_keys=True, indent=2)
        handle.write("\n")
        handle.flush()
        os.fsync(handle.fileno())
    try:
        os.replace(temporary, output)
    finally:
        temporary.unlink(missing_ok=True)
    print(f"artifact identity verified: {output}")
    return 0


if __name__ == "__main__":
    try:
        raise SystemExit(main())
    except Exception as exc:
        print(f"artifact identity failed: {exc}", file=sys.stderr)
        raise SystemExit(1)

File: tests/artifact-identity.mjs
#!/usr/bin/env node
import fs from 'node:fs';
import {spawnSync} from 'node:child_process';
import {resolve} from 'node:path';

const repo = resolve(new URL('..', import.meta.url).pathname);
const reporter = resolve(repo, 'scripts/artifact_identity.py');
let failed = false;
const fail = message => { console.error(`FAIL: ${message}`); failed = true; };

if (!fs.existsSync(reporter)) {
  fail('artifact identity reporter is missing');
} else {
  const work = fs.mkdtempSync('/extra/tmp/peers-artifact-test-');
  try {
    const uiRoot = resolve(work, 'ui-package');
    const coreRoot = resolve(work, 'core-package');
    for (const root of [uiRoot, coreRoot]) {
      fs.mkdirSync(resolve(root, 'variants/linux-amd64'), {recursive: true});
    }
    fs.writeFileSync(resolve(uiRoot, 'manifest.json'), '{"name":"peers_ui"}\n');
    fs.writeFileSync(resolve(uiRoot, 'variants/linux-amd64/PeersView.qml'), 'Item {}\n');
    fs.writeFileSync(resolve(coreRoot, 'manifest.json'), '{"name":"peers_core"}\n');
    fs.writeFileSync(resolve(coreRoot, 'variants/linux-amd64/peers_core_plugin.so'), 'core-bytes\n');

    const uiLgx = resolve(work, 'peers-ui.lgx');
    const coreLgx = resolve(work, 'peers-core.lgx');
    for (const [archive, root] of [[uiLgx, uiRoot], [coreLgx, coreRoot]]) {
      const tar = spawnSync('tar', ['czf', archive, '-C', root, '.'], {encoding: 'utf8'});
      if (tar.status !== 0) fail(`fixture archive failed: ${tar.stderr.trim()}`);
    }

    const iso = resolve(work, 'iso');
    const gui = resolve(iso, 'data/Logos/LogosBasecamp');
    const installedUi = resolve(gui, 'plugins/peers_ui');
    const installedCore = resolve(gui, 'modules/peers_core');
    fs.mkdirSync(installedUi, {recursive: true});
    fs.mkdirSync(installedCore, {recursive: true});
    fs.cpSync(resolve(uiRoot, 'variants/linux-amd64'), installedUi, {recursive: true});
    fs.cpSync(resolve(coreRoot, 'variants/linux-amd64'), installedCore, {recursive: true});
    fs.copyFileSync(resolve(uiRoot, 'manifest.json'), resolve(installedUi, 'manifest.json'));
    fs.copyFileSync(resolve(coreRoot, 'manifest.json'), resolve(installedCore, 'manifest.json'));
    fs.writeFileSync(resolve(installedUi, 'variant'), 'linux-amd64');
    fs.writeFileSync(resolve(installedCore, 'variant'), 'linux-amd64');
    fs.mkdirSync(resolve(installedUi, 'media-cache'));
    fs.writeFileSync(resolve(installedUi, 'media-cache/private-runtime-object'), 'not release payload\n');
    const appImage = resolve(work, 'Basecamp.AppImage');
    fs.writeFileSync(appImage, 'basecamp\n');
    const output = resolve(work, 'identity.json');
    const coreRev = '1'.repeat(40);
    const args = [reporter, '--ui-lgx', uiLgx, '--core-lgx', coreLgx,
      '--iso', iso, '--appimage', appImage, '--core-rev', coreRev, '--output', output];

    const good = spawnSync('python3', args, {encoding: 'utf8'});
    if (good.status !== 0) fail(`matching payload rejected: ${good.stderr.trim()}`);
    if (!fs.existsSync(output)) fail('identity report was not written');
    else {
      const report = JSON.parse(fs.readFileSync(output, 'utf8'));
      if (report.schema !== 1 || report.peers_core_revision !== coreRev)
        fail('identity report schema or pinned revision is wrong');
      for (const key of ['basecamp_appimage', 'peers_ui_lgx', 'peers_core_lgx']) {
        if (!/^[0-9a-f]{64}$/.test(report.artifacts?.[key]?.sha256 ?? ''))
          fail(`${key} SHA-256 is missing`);
      }
      if (report.components?.peers_ui?.file_count !== 3
          || report.components?.peers_core?.file_count !== 3)
        fail('installed payload file counts are wrong');
      if (JSON.stringify(report).includes('private-runtime-object'))
        fail('runtime cache contents leaked into the identity report');
    }

    const previous = fs.readFileSync(output, 'utf8');
    fs.writeFileSync(resolve(installedUi, 'PeersView.qml'), 'changed\n');
    const changed = spawnSync('python3', args, {encoding: 'utf8'});
    if (changed.status === 0) fail('changed installed payload was accepted');
    if (fs.readFileSync(output, 'utf8') !== previous) fail('failed verification replaced the last good report');
    fs.writeFileSync(resolve(installedUi, 'PeersView.qml'), 'Item {}\n');

    fs.writeFileSync(resolve(installedCore, 'unexpected'), 'extra\n');
    const extra = spawnSync('python3', args, {encoding: 'utf8'});
    if (extra.status === 0) fail('extra installed payload was accepted');
    fs.rmSync(resolve(installedCore, 'unexpected'));

    fs.rmSync(resolve(installedCore, 'peers_core_plugin.so'));
    fs.symlinkSync('/etc/passwd', resolve(installedCore, 'peers_core_plugin.so'));
    const linked = spawnSync('python3', args, {encoding: 'utf8'});
    if (linked.status === 0) fail('symlinked installed payload was accepted');
  } finally {
    fs.rmSync(work, {recursive: true, force: true});
  }
}

if (failed) process.exit(1);
console.log('ok: final artifact identity is deterministic and matches installed payloads');
